rasterio/_env.pyx

```diff
@@ -68,11 +68,11 @@ cdef void log_error(CPLErr err_class, int err_no, const char* msg) with gil:
 # other platforms. Each calls log_error().
 IF UNAME_SYSNAME == "Windows":
     cdef void __stdcall logging_error_handler(CPLErr err_class, int err_no,
-                                              const char* msg):
+                                              const char* msg) with gil:
         log_error(err_class, err_no, msg)
 ELSE:
     cdef void logging_error_handler(CPLErr err_class, int err_no,
-                                    const char* msg):
+                                    const char* msg) with gil:
```

Comment:
@jdmcbr I used gdb & lldb (the latter ended up being easier) on a test script that isolated the test case in a bit more self-contained way.  Incidentally, along the way, I removed the separately installed Python 3.5, and installed via brew - and now I get useful error reporting from OSX in this case that includes the same stack trace:

```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff890d6f06 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff977d64ec pthread_kill + 90
2   libsystem_c.dylib             	0x00007fff8b7bb6df abort + 129
3   org.python.python             	0x000000010db8418e Py_FatalError + 439
4   org.python.python             	0x000000010db861a6 PyThreadState_Get + 29
5   _env.cpython-36m-darwin.so    	0x0000000113bb7419 __pyx_f_8rasterio_4_env_logging_error_handler + 57 (_env.c:1943)
6   libgdal.1.dylib               	0x000000010e5d046f CPLErrorV + 507
7   libgdal.1.dylib               	0x000000010e5d025f CPLError + 134
8   libgdal.1.dylib               	0x000000010e5a7ac2 GDALOpenInternal(GDALOpenInfo&, char const* const*) + 366
9   libgdal.1.dylib               	0x000000010e5a7920 GDALOpenInternal(char const*, GDALAccess, char const* const*) + 49
10  libgdal.1.dylib               	0x000000010e5a7cb8 GDALOpenShared + 190
11  _base.cpython-36m-darwin.so   	0x000000010e2edf94 __pyx_pw_8rasterio_5_base_11DatasetBase_5start + 804 (_base.c:4378)
12  org.python.python             	0x000000010daffc9e _PyCFunction_FastCallDict + 528
13  org.python.python             	0x000000010db672f1 call_function + 562
14  org.python.python             	0x000000010db64691 _PyEval_EvalFrameDefault + 26868
15  org.python.python             	0x000000010db67ace _PyEval_EvalCodeWithName + 1910
16  org.python.python             	0x000000010db681ac fast_function + 241
17  org.python.python             	0x000000010db672cb call_function + 524
18  org.python.python             	0x000000010db64691 _PyEval_EvalFrameDefault + 26868
19  org.python.python             	0x000000010db67ace _PyEval_EvalCodeWithName + 1910
20  org.python.python             	0x000000010db5dcf4 PyEval_EvalCode + 100
21  org.python.python             	0x000000010db87077 run_mod + 58
22  org.python.python             	0x000000010db8734e PyRun_FileExFlags + 178
23  org.python.python             	0x000000010db86a75 PyRun_SimpleFileExFlags + 668
24  org.python.python             	0x000000010db9ad03 Py_Main + 3575
25  org.python.python             	0x000000010dab9e19 0x10dab8000 + 7705
26  libdyld.dylib                 	0x00007fff8cd1b5ad start + 1
```

In this case, this line `__pyx_f_8rasterio_4_env_logging_error_handler + 57 (_env.c:1943)` was a good hint that sent me looking into the C file generated by Cython.  From there, the comments embedded in that file sent me looking into the appropriate place in the Cython code.  At that point, it was trial and error to see what would work.  If you are able to get gdb working for you, it would be interesting to see your stack trace.